From a330ed7f2546ea29eb3116a1bb38ec4076c3d24e Mon Sep 17 00:00:00 2001
From: Krzysztof Parzyszek <Krzysztof.Parzyszek@amd.com>
Date: Fri, 16 Oct 2026 07:08:50 -0500
Subject: [PATCH] [flang][OpenMP] Ignore DECLARE_SIMD and warn about missing
 implementation

DECLARE_SIMD is a hint to the compiler and ignoring it is technically
speaking a conforming implementation (although obviously not desirable).
Do not abort compilation when DECLARE_SIMD is encountered, but emit
a warning that it's currently ignored.
---
 flang/lib/Semantics/check-omp-structure.cpp          |  4 ++++
 flang/test/Semantics/OpenMP/cray-pointer-usage.f90   |  1 +
 .../Semantics/OpenMP/declarative-directive01.f90     |  2 ++
 flang/test/Semantics/OpenMP/declare-simd-empty.f90   |  2 ++
 .../Semantics/OpenMP/declare-simd-interface-body.f90 |  6 +++++-
 flang/test/Semantics/OpenMP/declare-simd-scope.f90   | 12 ++++++++++++
 flang/test/Semantics/OpenMP/declare-simd-uniform.f90 |  2 ++
 flang/test/Semantics/OpenMP/declare-simd.f90         |  7 ++++++-
 .../Semantics/OpenMP/future-directive-spellings.f90  |  1 +
 flang/test/Semantics/OpenMP/linear-clause01.f90      |  9 +++++++++
 flang/test/Semantics/OpenMP/linear-clause02.f90      |  3 +++
 flang/test/Semantics/OpenMP/linear-clause03.f90      |  1 +
 flang/test/Semantics/OpenMP/simd-linear-array.f90    |  1 +
 .../Dialect/OpenMP/OpenMPToLLVMIRTranslation.cpp     |  7 +++++++
 14 files changed, 56 insertions(+), 2 deletions(-)

diff --git a/lib/Semantics/check-omp-structure.cpp b/lib/Semantics/check-omp-structure.cpp
index 07c9a5222883c7..e5b0433ef0220b 100644
--- a/lib/Semantics/check-omp-structure.cpp
+++ b/lib/Semantics/check-omp-structure.cpp
@@ -1928,6 +1928,12 @@
 void OmpStructureChecker::Enter(const parser::OmpDeclareSimdDirective &x) {
   const parser::OmpDirectiveName &dirName{x.v.DirName()};
 
+  unsigned version{context_.langOptions().OpenMPVersion};
+
+  context_.Say(dirName.source,
+      "%s directive is not implemented yet and will be ignored"_warn_en_US,
+      GetUpperName(dirName.v, version));
+
   const Scope &containingScope = context_.FindScope(dirName.source);
   const Scope &progUnitScope = GetProgramUnitContaining(containingScope);
 
